libsrc/hyperion/Hyperion.cpp

```diff
@@ -87,8 +87,9 @@ void Hyperion::start()
 		_ledStringColorOrder.push_back(led.colorOrder);
 	}
 
+	// NOTE: Removed visiblePriorityChanged->update connection, as it results in a double write()
 	// connect Hyperion::update with Muxer visible priority changes as muxer updates independent
-	connect(&_muxer, &PriorityMuxer::visiblePriorityChanged, this, &Hyperion::update);
```

Comment:
As far as i know this should catch the following:
- When a priority timeout ends (consider static color)
- or user requests a priority switch the leds won't update

The old solution was a timer for Hyperion:update without further conditions

Reply:
@brindosch and others

OK. Let's figure out what happens with or without connect.
You may need to help me with additional test-cases...., esp. for 
> or user requests a priority switch the leds won't update

What are the steps to execute that testcase?

### Test

All tests with 
- a "file" device with timestamps on

- Smoothing off (to see the pure update() calls)

**With Connect (PriorityMuxer::visiblePriorityChanged -> Hyperion::update)**

a) Start of hyperion

tail: output.test: file truncated
2020-02-09 16:02:50.554 | +0024 [{67,253,33}{255,196,0}{193,0,228}{0,199,241}]
**2020-02-09 16:02:50.555 | +0000 [{67,253,33}{255,196,0}{193,0,228}{0,199,241}]**
2020-02-09 16:02:51.555 | +1000 [{70,253,31}{253,192,1}{190,0,229}{0,202,240}]

-> Duplicate record at the same time (+0000 is the delta since last Led write)

b) When a priority timeout ends (consider static color)
Set an Effect-Color (Red = rgb(255,0,0))

...
2020-02-09 16:09:33.399 | +0999 [{191,188,188}{207,207,207}{207,202,203}{179,166,173}]
2020-02-09 16:09:33.424 | +0024 [{255,0,0}{255,0,0}{255,0,0}{255,0,0}]
**2020-02-09 16:09:33.424 | +0000 [{255,0,0}{255,0,0}{255,0,0}{255,0,0}]**
2020-02-09 16:09:43.647 | +10223 [{189,185,185}{207,207,207}{128,110,122}{86,63,79}]

-> Duplicate record at the same time (+0000 is the delta since last Led write)

**Without Connect (PriorityMuxer::visiblePriorityChanged -> Hyperion::update)**

a) Start of hyperion

tail: output.test: file truncated
2020-02-09 16:12:15.143 | +0037 [{67,253,33}{255,196,0}{193,0,228}{0,199,241}]
2020-02-09 16:12:16.143 | +1000 [{70,253,31}{253,192,1}{190,0,229}{0,202,240}]

-> No Duplicate records

b) When a priority timeout ends (consider static color)
Set an Effect-Color (Red = rgb(255,0,0))

...
2020-02-09 16:13:31.899 | +0999 [{1&PriorityMuxer::visiblePriorityChanged, _ledDevice, &LedDevice::visiblePriorityChanged,82,178,179}{207,207,207}{214,215,216}{186,178,185}]
2020-02-09 16:13:32.722 | +0823 [{255,0,0}{255,0,0}{255,0,0}{255,0,0}]
2020-02-09 16:13:43.899 | +11176 [{183,179,180}{207,207,207}{128,110,122}{85,60,77}]

-> No Duplicate records

### Additional questions/findings

1. Are there other cases/scenarios that should be tested?
2. There is a connect in the DeviceWrapper (PriorityMuxer::visiblePriorityChanged ->&LedDevice::visiblePriorityChanged) which seems not relevant.
I suggest to remove that too, removing the tight coupling between components.

Reply:
OK. Seems I found the scenario where no output is generated, if the connect is missing.

a) Set a fixed colour via the Systray as a new source.
-> Colour is written to the output
b) Switch to Platform Capture source
-> Output is updated
c) Switch to Fixed Colour source
-> No update is written
 
The duplicate writes in the other scenarios are not necessarily bad, but maybe there is a better alternative avoiding them. Let me have a look.

Reply:
We need lot's of tests that can run at ci.

A test should be:
- Set color at prio 1, set another color at prio 2, select prio 1 [Hyperion::setCurrentSourcePriority](https://github.com/hyperion-project/hyperion.ng/blob/master/libsrc/hyperion/Hyperion.cpp#L286)
No one calls Hyperion::update()
It should be possible to call update() on success like [Hyperion::setSourceAutoSelectEnabled](https://github.com/hyperion-project/hyperion.ng/blob/master/libsrc/hyperion/Hyperion.cpp#L280)
- Same with [Hyperion::clear()](https://github.com/hyperion-project/hyperion.ng/blob/master/libsrc/hyperion/Hyperion.cpp#L394)
- Last case: Timeout, a priority is dropped and therefore another priority is now active, who updates?

Reply:
@brindosch Would you mind outlining how I would do 

> Set color at prio 1, set another color at prio 2, select prio 1

via the UI? I only have the capture source and everything else gets prio 1 assigned.

Or are you describing just automatic test-cases? 

Additional questions:

When I manually set an Effect-Color or Picture Streaming via the Browser, both get always prio 1 and refresh with 10hz. Is it possible to change the refresh rate or is it always defaulted?
In this context, the UI states "Duration 2s", but the streaming does not end after 2s. What does the "2s" statement mean?

Reply:
>> Set color at prio 1, set another color at prio 2, select prio 1

> via the UI? I only have the capture source and everything else gets prio 1 assigned.

Not possible, as the prio can't be configured at the ui

> When I manually set an Effect-Color or Picture Streaming via the Browser, both get always prio 1 and refresh with 10hz. Is it possible to change the refresh rate or is it always defaulted?

Is defaulted, the screen capture consumes lot's of memory. I still need to figure out the bottleneck here.

> In this context, the UI states "Duration 2s", but the streaming does not end after 2s. What does the "2s" statement mean?

It reflects the timeout from PriorityMuxer (which is set for each single image). This is visual hidden für other implementations (proto/flat). We may need to distinguish between an "image" and an "imagestream". Especially as proto/flat is just a protocol and nothing "special". A cleanup of the components/PriorityMuxer to outline the belongings better is required. But it's not always easy. An effect for example can be a "ledColorsStream" but also an "imagestream"

Reply:
> It reflects the timeout from PriorityMuxer (which is set for each single image). 

The current "Duration 2s" on the UI created a different expectation to myself as a user rather than a timeout. Maybe the text is better removed from the screen to avoid confusion?

Reply:
Yes, the problem is that you can set a single image with timeout, and this timeout will be lost, as both are going through the same api method and there is no flag to mark it accordingly.

Reply:
> Yes, the problem is that you can set a single image with timeout, and this timeout will be lost, as both are going through the same api method and there is no flag to mark it accordingly.

Maybe I did not express myself clearly...

As a hyperion user, I do not find the duration information rather confusing than helpful:
https://snipboard.io/du20Jw.jpg

It gives too much room for interpretation what the time value may be.
I was just looking from a usability perspective, not with my developer lenses on. :)